libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java

```diff
@@ -247,6 +247,8 @@ protected int readFromChannel(InboundChannelBuffer channelBuffer) throws IOExcep
         if (bytesRead < 0) {
             closeNow = true;
             return 0;
+        } else if (bytesRead == 0) {
```

Comment:
Small optimization here to make up for the fact that this path would be take a lot more often with this change on accepts.